File: 33_rabin_karp.cpp
#include <bits/stdc++.h>
using namespace std;

/*
Task 33: Implement the Rabin–Karp string matching algorithm.
Reads two lines: text and pattern; prints starting indices separated by spaces.
*/

int main()
{
    string t, p;
    if (!getline(cin, t))
        return 0;
    if (!getline(cin, p))
        return 0;
    int n = t.size(), m = p.size();
    if (m > n)
    {
        cout << '\n';
        return 0;
    }
    const unsigned long long base = 1315423911ULL;
    unsigned long long hashP = 0, hashT = 0, powB = 1;
    for (int i = 0; i < m; ++i)
    {
        hashP = hashP * 131 + (unsigned char)p[i];
        hashT = hashT * 131 + (unsigned char)t[i];
        if (i)
            powB *= 131;
    }
    vector<int> res;
    if (hashP == hashT && t.substr(0, m) == p)
        res.push_back(0);
    for (int i = m; i < n; ++i)
    {
        hashT = hashT * 131 + (unsigned char)t[i] - powB * (unsigned char)t[i - m];
        if (hashT == hashP && t.substr(i - m + 1, m) == p)
            res.push_back(i - m + 1);
    }
    for (size_t i = 0; i < res.size(); ++i)
    {
        if (i)
            cout << ' ';
        cout << res[i];
    }
    cout << '\n';
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(1)

File: 33_rabin_karp.py
"""
Task 33: Implement the Rabin–Karp string matching algorithm.
Reads two lines: text and pattern; prints starting indices separated by spaces.
"""
import sys

def rabin_karp(t: str, p: str):
    n, m = len(t), len(p)
    if m > n: return []
    base = 256
    mod = 2**61-1
    def mul(a,b): return (a*b) % mod
    hp = 0; ht = 0; power = 1
    for i in range(m):
        hp = (hp*base + ord(p[i])) % mod
        ht = (ht*base + ord(t[i])) % mod
        if i>0: power = (power*base) % mod
    res = []
    if hp == ht and t[:m] == p: res.append(0)
    for i in range(m, n):
        ht = (ht*base + ord(t[i]) - power*ord(t[i-m])) % mod
        if ht < 0: ht += mod
        if ht == hp and t[i-m+1:i+1] == p: res.append(i-m+1)
    return res

if __name__ == '__main__':
    t = sys.stdin.readline().rstrip('\n')
    p = sys.stdin.readline().rstrip('\n')
    print(' '.join(map(str, rabin_karp(t,p))))

# Time Complexity: O(n)
# Space Complexity: O(1)
